public/app/plugins/panel/flamegraph/components/FlameGraph/rendering.ts

```diff
@@ -0,0 +1,134 @@
+import { colors, fuzzyMatch } from '@grafana/ui';
+
+import {
+  BAR_BORDER_WIDTH,
+  BAR_TEXT_PADDING_LEFT,
+  COLLAPSE_THRESHOLD,
+  HIDE_THRESHOLD,
+  LABEL_THRESHOLD,
+  PIXELS_PER_LEVEL,
+} from '../../constants';
+
+import { ItemWithStart } from './dataTransform';
+
+type RectData = {
+  width: number;
+  height: number;
+  x: number;
+  y: number;
+  collapsed: boolean;
+  ticks: number;
+  label: string;
+};
+
+/**
+ * Compute the pixel coordinates for each bar in a level. We need full level of bars so that we can collapse small bars
+ * into bigger rects.
+ * @param level
+ * @param levelIndex
+ * @param totalTicks
+ * @param rangeMin
+ * @param pixelsPerTick
+ */
+export function getRectDimensionsForLevel(
+  level: ItemWithStart[],
+  levelIndex: number,
+  totalTicks: number,
+  rangeMin: number,
+  pixelsPerTick: number
+): RectData[] {
+  const coordinatesLevel = [];
+  for (let barIndex = 0; barIndex < level.length; barIndex += 1) {
+    const item = level[barIndex];
+    const barX = getBarX(item.start, totalTicks, rangeMin, pixelsPerTick);
+    let curBarTicks = item.value;
+
+    // merge very small blocks into big "collapsed" ones for performance
+    const collapsed = curBarTicks * pixelsPerTick <= COLLAPSE_THRESHOLD;
+    if (collapsed) {
+      while (
+        barIndex < level.length - 1 &&
+        item.start + curBarTicks === level[barIndex + 1].start &&
+        level[barIndex + 1].value * pixelsPerTick <= COLLAPSE_THRESHOLD
+      ) {
+        barIndex += 1;
+        curBarTicks += level[barIndex].value;
+      }
+    }
+
+    const width = curBarTicks * pixelsPerTick - (collapsed ? 0 : BAR_BORDER_WIDTH * 2);
+    coordinatesLevel.push({
+      width,
+      height: PIXELS_PER_LEVEL,
+      x: barX,
+      y: levelIndex * PIXELS_PER_LEVEL,
+      collapsed,
+      ticks: curBarTicks,
+      label: item.label,
+    });
+  }
+  return coordinatesLevel;
+}
+
+export function renderRect(
+  ctx: CanvasRenderingContext2D,
+  rect: RectData,
+  totalTicks: number,
+  rangeMin: number,
+  rangeMax: number,
+  query: string,
+  levelIndex: number,
+  topLevelIndex: number
+) {
+  if (rect.width < HIDE_THRESHOLD) {
+    return;
+  }
+
+  ctx.beginPath();
+  ctx.rect(rect.x + (rect.collapsed ? 0 : BAR_BORDER_WIDTH), rect.y, rect.width, rect.height);
+
+  //  / (rangeMax - rangeMin) here so when you click a bar it will adjust the top (clicked)bar to the most 'intense' color
+  const intensity = Math.min(1, rect.ticks / totalTicks / (rangeMax - rangeMin));
+  const h = 50 - 50 * intensity;
+  const l = 65 + 7 * intensity;
+
+  const name = rect.label;
+  const queryResult = query && fuzzyMatch(name.toLowerCase(), query.toLowerCase()).found;
```

Comment:
I feel like with a search this lenient we may want to bold the highlighted ranges fuzzyMatch returns so people understand why things are being highlighted.

Reply:
100% - this is something we discussed during the week. I will be updating that next week :)